#166238094 Users should receive a JWT on Successful Registration/Login

ayodejiAA opened the pull request:
#### What does this PR do?
- [x] Send token to users upon successful registration or login

#### Description of Task to be completed?
- [x] Have authentication working on the following routes- 
```/api/v1/auth/signup``` and ```/api/v1/auth/login```

#### How should this be manually tested?
This feature can be tested with  any API testing application, like postman. It can as well be tested on the swagger documentation route. 
- Clone the REPO and checkout to ft-implement-jwt-authentication-166238094. Add all required env variables. 
- Run ```npm run buildTables``` follow by ```npm start```, and use your API testing app to test the above API endpoints. 
>- Required fields for signup are ```userName, email, firstName, lastName and password```
>- Required fields for login are ```userLogin (accepts email/userName) and password```

You can also point your browser to ```/api-docs/``` route to test the feature on swagger API documentation. 

#### Any background context you want to provide?
Authentication has always been an important aspect of every application. During registration and login, user should be provided token that will be required for authentication and use on subsequent calls to other application endpoints. 

#### What are the relevant pivotal tracker stories?
[#166238094](https://www.pivotaltracker.com/story/show/166238094)

#### Screenshots (if appropriate)
---
![sam1](https://user-images.githubusercontent.com/42908597/58758562-da4f9700-8514-11e9-8f94-b37a1f2e8366.PNG)

abdulfatai360 commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "index.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha ./server/test --require @babel/register --colors --slow 400 --timeout 5000 --exit",
+    "test": "npm run clean-tables && cross-env NODE_ENV=test nyc --reporter=text mocha ./server/test/*.spec.js --require @babel/register --require @babel/polyfill --slow 400 --timeout 5000 --exit",
```
Bro, well done for adding the `babel polyfill` plugin to the test script. However, if I may ask, why did you remove the `--colors` option?

ayodejiAA replied:
Thanks. I  might have missed that when resolving conflicts. I will add it in my next commit.

Xwebyna replied:
Thanks. I  might have missed when resolving conflicts. I will add it in my next commit.

abdulfatai360 commented on package.json:
```diff
@@ -28,8 +29,10 @@
   "homepage": "https://github.com/los-deferal-02/Hello-Books#readme",
   "dependencies": {
     "@babel/polyfill": "^7.4.4",
+    "bcrypt": "^3.0.6",
```
Bro, sorry. Bcrypt doesn't work on my PC. I have issues with its dependencies. Could you please consider using `bcryptjs` npm package instead?

ayodejiAA replied:
Okay bro. I will.

abdulfatai360 commented on server/controllers/userController.js:
```diff
@@ -0,0 +1,91 @@
+      }
+      const passwordValid = await decryptPassword(password, user.password);
+      if (!passwordValid) {
+        return res.status(401).json({
```
Your HTTP response specification does not follow the team's agreed spec. You may refer to this [https://github.com/omniti-labs/jsend](https://github.com/omniti-labs/jsend) for guidance

abdulfatai360 commented on server/test/root-route.spec.js:
```diff
@@ -5,16 +5,15 @@ import app from '../../index';
+describe('Root url Test', () => {
   describe('App Root', () => {
-    it('should return status 200 and a message attribute', (done) => {
+    it('should return status 404 and a message attribute', (done) => {
```
Bro, well done for the job done so far. But what is the idea behind the `root URL` sending a `404 - Not found` error?

abdulfatai360 reviewed (changes requested):
@Xwebyna  Awesome efforts in implementing the signin and signup feature. However, I raised some salient points that needs to be addressed. Please take a look at them.

tolulope-od commented on hello-books-swagger.json:
```diff
@@ -1,97 +1,162 @@
 {
```
Great work with the controller @Xwebyna I especially like how you've added an error handling middleware to the application. However, could you please let us know why you deleted the old setup of the `swaggerDoc`?

tolulope-od commented on server/middlewares/errors.js:
```diff
@@ -0,0 +1,28 @@
+  */
+const errorHandler = (err, req, res, next) => {
+  res.status(500).json({
+    status: res.statusCode,
```
Great work here Ayodeji! I like how you've abstracted the error handling logic. Could you please check however that the response specification is according to what was agreed on for the [JSend](https://github.com/omniti-labs/jsendhttps://github.com/omniti-labs/jsend) specification for server errors? 

Once again, this is fantastic stuff!

ayodejiAA replied:
Thanks for pointing it out, Tolu.  Necessary adjustments will be made, including updating the swagger doc. Thanks.

chialuka commented on server/controllers/userController.js:
```diff
@@ -0,0 +1,91 @@
+
+      return res.status(201).json({
+        status: res.statusCode,
+        token
```
This is massive, @Xwebyna and it is so well done. Thumbs up.

I'd like to know, if all we return is a token on signup, there'll be a need to make a fresh request on the front-end if the user is being redirected tho their dashboard, won't there?

To avoid this situation, won't it be better to return the details of the user except the password?

Xwebyna replied:
ThanI thought of that as well

ayodejiAA replied:
Thanks @LukasChiama. I thought of that as well, but then the story only instructs to implement jwt authentication on the app. So, there will be a fresh request to get the complete profile details of the user when opening the dashboard. Moreover, the frontend will be built with ReactJs which tends to make this easier.

chialuka reviewed (commented):
Great job @Xwebyna. This is massive.

chiboycalix reviewed (approved):
Nice job @Xwebyna, your work is well documented in your PR and it's easy to understand. Thumbs up

tolulope-od commented on hello-books-swagger.json:
```diff
@@ -1,97 +1,200 @@
+						"description": "This is the request body object containing users' sign in information. UserLogin can be email or username",
+						"required": true,
+						"schema": {
+							"$ref": "#/definitions/UserLoginRequest"
```
Great work here Ayodeji, thanks for reverting some of the changes on this file. I do have a few concerns here though, I feel that the removal of the User definition doesn't exactly serve the purpose of the documentation. I also don't feel so much confidence about having definitions for login and signup requests. They could just be added as examples in the path method documentation, while definitions would be reserved for our models in the application such as a `User`, `Book` and possibly our `apiResponse` format. Let me know what you think about this.

ayodejiAA replied:
Thanks for the comment @tolulope-od. I agree with your point. The model should contain the db models not input field definitions.

tolulope-od commented on hello-books-swagger.json:
```diff
@@ -1,97 +1,200 @@
+			"url": "http://www.apache.org/licenses/LICENSE-2.0.html"
+		}
+	},
+	"host": "localhost:4000",
```
Hey @Xwebyna did you know that setting the host to `localhost:4000` could cause the Documentation to fail when it's hosted on Heroku?

ayodejiAA replied:
Yeah, I added it for development purpose.

Nerocodes commented on server/config/buildTables.js:
```diff
@@ -1,14 +1,18 @@
-    user_name VARCHAR(100) NOT NULL,
+    username VARCHAR(100) UNIQUE NOT NULL,
+    firstname VARCHAR(100) NOT NULL,
+    lastname VARCHAR(100) NOT NULL,
```
I'm not sure we decided on a naming convention for our db columns but I think underscores are more readable for serperating words. Are there any particular reasons you chose this?

ayodejiAA replied:
@Nerocodes Did you see the dB diagram we created? I believe we all agreed on that at the beginning of the sprint.

Nerocodes replied:
Alright!

Nerocodes reviewed (approved):
I left a comment about the naming convention we are using, asides that everything seems good. Great job!

abdulfatai360 reviewed (approved):
Nice work @Xwebyna

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
Fix test issue

ayodejiAA committed:
feature(authentication): User should receive JWT---
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
feature(authentication): User should receive JWT..:
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger
